hibernate-core/src/main/java/org/hibernate/sql/results/graph/collection/internal/AbstractCollectionInitializer.java

```diff
@@ -85,6 +91,69 @@ public void resolveKey(RowProcessingState rowProcessingState) {
 		}
 	}
 
+	protected void resolveInstance(RowProcessingState rowProcessingState, boolean isEager) {
+		if ( collectionKey != null ) {
+			final SharedSessionContractImplementor session = rowProcessingState.getSession();
+			final PersistenceContext persistenceContext = session.getPersistenceContext();
+
+			final LoadingCollectionEntry loadingEntry = persistenceContext.getLoadContexts()
+					.findLoadingCollectionEntry( collectionKey );
+
+			if ( loadingEntry != null ) {
+				collectionInstance = loadingEntry.getCollectionInstance();
+				return;
+			}
+
+			final PersistentCollection<?> existing = persistenceContext.getCollection( collectionKey );
+
+			if ( existing != null ) {
+				collectionInstance = existing;
+				return;
+			}
+
+			final CollectionPersister collectionDescriptor = collectionAttributeMapping.getCollectionDescriptor();
+			final CollectionSemantics<?, ?> collectionSemantics = collectionDescriptor.getCollectionSemantics();
+			final Object key = collectionKey.getKey();
+
+			persistenceContext.forEachCollectionEntry(
+					(persistentCollection, collectionEntry) -> {
+						if ( collectionEntry.getCurrentPersister() == collectionDescriptor &&
+								key.equals( collectionEntry.getCurrentKey() ) ) {
+							collectionInstance = persistentCollection;
+						}
+					}, true
+			);
```

Comment:
I'm not sure I understand how this code can be relevant. AFAIU, this code should be doing the same as `persistenceContext.getCollection( collectionKey );` on line 107, but if there were an existing collection, we wouldn't even reach this code I marked. Can you explain, maybe also as Java comment, how this special condition can occur?

Reply:
yeap

Reply:
It's delayed insertion of the collection declarer, right?  Assuming so, we should look to circumvent this when the owner is not using IDENTITY generation and/or delayed insertion is by-passed

Reply:
> while for IDENTITY the autoflush is not executed before the query.

Right.  That's the delayed insertion.


> Basically we have to be sure the insert has not yet happened, but not sure how to check this

IIRC the generated-id has some well-defined form and you should be able to tell from the value itself.  Been a long time since I wrote all that though and maybe I misremember

Reply:
To be clear, this is on the EntityKey and/or EntityEntry, not the entity itself.

BTW, you should be able to easily test this theory of delayed insertion being the culprit - change from `#persist` to `#save`.  Because the latter is expected to return the generated id, we are not able to delay the insert there.  But otherwise, the behavior would be the same

Reply:
> I debugged a little more and what I noticed id that with an assigned ID value before executing the query an autoflush is executed that for all `persistencecontext.collectionEntries` calls a `persistenceContext.addCollectionByKey` while for IDENTITY the autoflush is not executed before the query.

I traced the same root cause; then I further found the root cause is the "early insertion" in ActionQueue leaves no trace of  the table to be updated. As long as we fixed that, the flush would happen for Identity type. If you were interested, you could refer to my PR closed for details.

Reply:
paired with @beikov and reworked the solution.

@NathanQingyangXu the method you changed I think is used to tell if there are some data to be sync with the db in this case there are not, the insert had already happened.

Reply:
> paired with @beikov and reworked the solution.
> 
> @NathanQingyangXu the method you changed I think is used to tell if there are some data to be sync with the db in this case there are not, the insert had already happened.

thanks for letting me know. obviously this bug is too complex to me, :(.